App Sec Custom Rules (#13109)
DataDog/documentation: content/en/security_platform/application_security/custom_rules.md

In the “Notify” section, configure zero or more [notification targets][4] for each rule case.

You can also create [notification rules][5] to eleviate manual edits to notification preferences for individual detection rules.

### Time windows

An `evaluation window` is specified to match when at least one of the cases matches true. This is a sliding window and evaluates in real-time.

Once a signal is generated, the signal will remain “open” if a case is matched at least once within this `keep alive` window. Each time a new event matches any of the cases, the *last updated* timestamp is updated for the signal.

A signal will “close” regardless of the query being matched once the time exceeds the `maximum signal duration`. This time is calculated from the first seen timestamp.

Additional cases can be added by clicking the **Add Case** button.

**Note**: The `evaluation window` must be less than or equal to the `keep alive` and `maximum signal duration`.

## Say what's happening

The **Rule name** section allows you to configure the rule name that appears in the rules list view, as well as the title of the signal.

The notification box has the same Markdown and preview features.

> ### Template variables

Author: ```suggestion
#### Template variables
```

---
title: Custom Detection Rules
kind: documentation
further_reading:
- link: "/security_platform/application_security/"
  tag: "Documentation"
  text: "Monitoring threats with Datadog Application Security"
- link: "/security_platform/application_security/troubleshooting"
  tag: "Documentation"
  text: "Troubleshoot common Datadog Application Security issues"
---

## Overview

Application Security comes with a set of [out-of-the-box detection rules][1] which aim to catch attack attempts and vulnerability triggers that impact your production systems.

> However, there are situations where you may want to customize a rule based on your environment. For example, you may want to customize a rule that catches attack attempts on a service’s route. For instance, a pre-production development route accepting SQL and returning the results. Catching SQL attempts is noisy as the route is restricted to internal developers, therefore you can customize this rule to exclude these patterns.

Reviewer: However, there are situations where you may want to customize a **detection** rule...

> However, there are situations where you may want to customize a rule based on your environment. For example, you may want to customize a rule that catches attack attempts on a service’s route. For instance, a pre-production development route accepting SQL and returning the results. Catching SQL attempts is noisy as the route is restricted to internal developers, therefore you can customize this rule to exclude these patterns.

Author: ```suggestion
However, there are situations where you may want to customize a rule based on your environment. For example, you may want to customize a detection rule that catches attack attempts on a service’s route. For instance, a pre-production development route accepting SQL and returning the results. Catching SQL attempts is noisy as the route is restricted to internal developers, therefore you can customize this rule to exclude these patterns.
```

Another example is customizing a rule to exclude an internal security scanner. Application Security detects its activity as expected. However, you may not want to be notified of it’s regularly occurring scan.

In these situations, a custom detection rule can be created to exclude such events. This guide shows you how to create a custom detection rule for Application Security.

## Configuration

To customize an OOTB detection rule, you must first clone an existing rule. Navigate to your [Detection Rules][2] and select a rule. Scroll to the bottom of the rule and click the Clone Rule button. This now enables you to edit the existing rule.

### Define AppSec query

> Construct an Application Security query. For example, create a query to monitor an endpoint for SQL injection attempts: `@appsec.type:sql_injection @http.url_details.path:"/debug-endpoint-executing-sql" env:production`.

Reviewer: Sorry that's my bad, I thought the query would be used in the context of an Exclusion Filter. If it is a detection rule query, then it misses an `-`: `@appsec.type:sql_injection -@http.url_details.path:"/debug-endpoint-executing-sql" env:production`

> Construct an Application Security query. For example, create a query to monitor an endpoint for SQL injection attempts: `@appsec.type:sql_injection @http.url_details.path:"/debug-endpoint-executing-sql" env:production`.

Author: ```suggestion
Construct an Application Security query. For example, create a query to monitor an endpoint for SQL injection attempts: `@appsec.type:sql_injection -@http.url_details.path:"/debug-endpoint-executing-sql" env:production`.
```

### Set a rule case

#### Trigger

Rule cases, such as `successful trigger > 0`, are evaluated as case statements. Thus, the first case to match generates the signal. Create one or multiple rule cases, and click on the grey area next to them to drag and manipulate their ordering.

A rule case contains logical operations (`>, >=, &&, ||`) to determine if a signal should be generated based on the event counts in the previously defined queries.

**Note**: The query label must precede the operator. For example, `a > 3` is allowed; `3 < a` is not allowed.

Provide a **name** for each rule case. This name is appended to the rule name when a signal is generated.

#### Severity and notification

Set the severity of the signal. The dropdown allows you to select an appropriate severity level (`INFO`, `LOW`, `MEDIUM`, `HIGH`, `CRITICAL`).

In the “Notify” section, configure zero or more [notification targets][3] for each rule case.

> You can also create [notification rules][4] to eleviate manual edits to notification preferences for individual detection rules.

Author: LOL, what a mix of evaluate + alleviate, but i'll stick with alleviate!

> You can also create [notification rules][4] to eleviate manual edits to notification preferences for individual detection rules.

Author: ```suggestion
You can also create [notification rules][4] to alleviate manual edits to notification preferences for individual detection rules.
```

### Time windows

An `evaluation window` is specified to match when at least one of the cases matches true. This is a sliding window and evaluates in real-time.

> Once a signal is generated, the signal will remain “open” if a case is matched at least once within this `keep alive` window. Each time a new event matches any of the cases, the *last updated* timestamp is updated for the signal.

Reviewer 2: ```suggestion
When a signal is generated, it remains "open" if a case is matched at least once within this `keep alive` window. Each time a new event matches any of the cases, the *last updated* timestamp is updated for the signal.
```

> A signal will “close” regardless of the query being matched once the time exceeds the `maximum signal duration`. This time is calculated from the first seen timestamp.

Reviewer 2: ```suggestion
A signal will "close" regardless of the query being matched once the time exceeds the `maximum signal duration`. This time is calculated from the first seen timestamp.
```

Additional cases can be added by clicking the **Add Case** button.

**Note**: The `evaluation window` must be less than or equal to the `keep alive` and `maximum signal duration`.

### Say what's happening

The **Rule name** section allows you to configure the rule name that appears in the rules list view, as well as the title of the signal.

The notification box has the same Markdown and preview features.

#### Template variables

> Detection rules support template variables within the markdown notification box. Template variables permit injection of dynamic context from traces directly into a security signal and its associated notifications.

Reviewer 2: ```suggestion
Detection rules support template variables within the Markdown notification box. Template variables permit injection of dynamic context from traces directly into a security signal and its associated notifications.
```

```text
{{eval "first_seen_epoch-15*60*1000"}}
```

The attributes can be seen on a signal in the JSON dropdown, and you can access the attributes with the following syntax: `{{@attribute}}`. You can access inner keys of the event attributes by using JSON dot notation (for example, `{{@attribute.inner_key}}`).

**Note**: You can copy the raw JSON directly from a security signal. Select any security signal in the Signals Explorer to view its details. Click the export button in the top left corner, and select **Copy raw JSON to clipboard**.

This JSON object is an example of event attributes which may be associated with a security signal:

```json
{
  "attributes":{
    "title":"Security scanner detected",
    "http":{
      "url":"http://www.example.com"
    },
    "rule":{
      "detectionMethod":"threshold",
      "name":"Your rule name"
    },
    "events_matched":2,
    "first_seen":"2022-01-26T13:23:33.000Z",
    "last_seen":"2022-01-27T04:01:57.000Z"
  },
  "groupByPaths":[
    "service"
  ]
}
```

> For this attribute, you would use the following attribute in the “say what’s happening” section:

Reviewer 2: ```suggestion
For this attribute, use the following in the “say what’s happening” section:
```

---
title: Writing Custom Detection Rules
kind: documentation
---

## Overview

> To create a custom detection rule in Datadog, hover over **Security**, select **Detection Rules**, and select the **New Rule** button in the top right corner of the page.

Author: Premise of custom rules goes here

Author: Main example goes here

Author: Primary use case: customizing OOTB rules

#### Joining queries

Joining queries to span a timeframe can increase the confidence or severity of the Security Signal. For example, to detect a successful attack, both successful and unsuccessful tiggers can be correlated for a service.

Queries are correlated together by using a `group by` value. The `group by` value is typically an entity (for example, `IP address` or `Service`), but can be any attribute. In this instance, joined queries technically hold same attribute value because the value must be the same for the case to be met. If a `group by` value doesn’t exist, the case will never be met. A Security Signal is generated for each unique `group by` value when a case is matched.

For example, create a query that searches for the same `Service` activity, but append opposing HTTP status code queries:

> Query `successful_attack_trigger`: `security_activity_to_monitor` ``-@http_status_code:[500-599]`.

Author: ```suggestion
Query `successful_attack_trigger`: `security_activity_to_monitor -@http_status_code:[500-599]`.
```

> Query `failed_attack_trigger`: `security_activity_to_monitor``@http_status_code:[500-599]`.

Author: ```suggestion
Query `failed_attack_trigger`: `security_activity_to_monitor @http_status_code:[500-599]`.
```

Set `group by` for both to `Service` and create rule cases for each.

{{% /tab %}}

{{% tab "Anomaly" %}}

Construct a new Application Security query.

Optionally, define a unique count and signal grouping. Count the number of unique values observed for an attribute in a given timeframe. The defined `group by` option generates a signal for each `group by` value. Typically, `group by` is an entity (like `IP` or `Service`).

Anomaly detection inspects how the `group by` attribute has behaved in the past. If a `group by` attribute is seen for the first time (for example, the first time an IP is communicating with your system) and is anomalous, it will not generate a security signal because the anomaly detection algorithm has no historical data to base its decision on.

{{% /tab %}}
{{< /tabs >}}

## Set a rule case

{{< tabs >}}
{{% tab "Threshold" %}}

### Trigger

> Rule cases, such as `succesful trigger > 0`, are evaluated as case statements. Thus, the first case to match generates the signal. Create one or multiple rule cases, and click on the grey area next to them to drag and manipulate their ordering.

Author: ```suggestion
Rule cases, such as `successful trigger > 0`, are evaluated as case statements. Thus, the first case to match generates the signal. Create one or multiple rule cases, and click on the grey area next to them to drag and manipulate their ordering.
```
